### ISSUE TITLE
Support lazy loading AMD plugin modules

### ISSUE THREAD

**Issue by whatitis1** [OpenNTF/dojo-webpack-plugin#191]
I have a Dojo project that uses AMD plugins that load resource dynamically. These resources are only available at runtime, not build time. At runtime, my plugin tries to load the resource asynchronously and since the loaderProxy expects the plugin to load synchronously, I get an error that indicates that the module is unavailable. To give an example of what I'm working with:

```javascript
define([
    "dojo/request/script"
], function(script) {
    var cache = {};
    var pending = {};

    return {
        load: function(id, req, callback) {
            if(id in cache) {
                callback(cache[id]);
                return;
            }
            else if(id in pending) {
                // Hold onto the callback so we can resolve the call once the module loads
                pending[id].push(callback);
                return;
            }

            pending[id] = [callback];

            var path = ...; // resolve id to a resource path
            script.get(path).then(function() {
                // Once the module is loaded into global space, get a handle on the module
                var module = ...;

                // Resolve pending module requests
                var allPending = pending[id];
                for(var i = 0; i < pendingList.length; i++) {
	            allPending[i](module);
                }

                delete pending[id];
            });
        }
    };
});
```
There are then many other modules that depend on this plugin. For example:
```javascript
define([
    "my/plugin!my-module"
], function(myModule) {
    // leverage the module directly
});
```

Is this something that can be supported? I'd be happy to help, just have no idea where to start given I'm new to webpack and the dojo-webpack-plugin.

**Comment by chuckdumont** [comment]
That's a much taller order.  Module loading in webpack is synchronous, except when explicitly async (i.e. async require).  You couldn't do it without replacing webpack's module loader with an async module loader.

**Comment by chuckdumont** [comment]
Thinking about this some more, it may not be so difficult to do.  dojo/proxyLoader would still need to call the load method synchronously, but if the resource load did not complete synchronously, it could pass a Promise to the data instead of the data itself.

You'd also need to modify the injected code that invokes the define callback so that the callback is called only after all promises in the dependency array have been resolved.  See [AMDDefineDependency.js](https://github.com/webpack/webpack/blob/master/lib/dependencies/AMDDefineDependency.js) for the injected code in webpack that would need to be updated.  I'd make async completion a special case to avoid introducing regressions.  For that reason, you'd probably want to use Dojo promises (i.e. dojo/Defered) rather than ES6 Promise because the later does not allow for synchronous interrogation of the resolved state.

**Comment by chuckdumont** [comment]
#195

**Comment by ddumontatibm** [comment]
To make the interface better, could you just always return a promise?

**Comment by chuckdumont** [comment]
Not sure of the difference between always returning a promise vs. always assuming that it can be a promise and using `Promise.resolve(...)`.  Always returning a promise would make the implementation more complicated and I don't really see the need.

**Comment by ddumontatibm** [comment]
I thought you might be calling load on a promise or a module.   You could just load the Promise.resolve() instead and hide the end user from having to guess.

**Comment by ddumontatibm** [comment]
ok

**Comment by chuckdumont** [comment]
@whatitis1, please try out 2.7.8-0.  See the [async](/OpenNTF/dojo-webpack-plugin#async) options for details.

**Comment by cmdcolin** [comment]
@chuckdumont I was trying this out for curiosity and saw it seemed to maybe give an error with dojo/request/xhr. I edited the dojo-webpack-plugin-sample like this and there is an error about util in xhr being undefined

```
Uncaught (in promise) TypeError: Cannot read property 'addCommonMethods' of undefined
    at xhr.js:386
    at xhr.js:389
```

Code

```
diff --git a/js/bootstrap.js b/js/bootstrap.js
index 63d7d91..8305673 100644
--- a/js/bootstrap.js
+++ b/js/bootstrap.js
@@ -15,14 +15,20 @@
  */
 define([
        "dojo/ready",
+       "dojo/request/xhr",^M
        "dojo/parser",
        "dijit/layout/TabContainer",
        "js/LazyContentPane",
        "dojo/has!webpack?dojo-webpack-plugin/amd/dojoES6Promise",
        "css!dijit/themes/claro/claro.css",
        "css!theme/sample.less"
-], function(ready, parser) {
+], function(ready, xhr, parser) {^M
+^M
        ready(function() {
+        xhr('http://google.com')^M
+            .then(response => console.log(response))^M
+            .catch(err => console.error(err))^M
+^M
                parser.parse();
        });
        return {};
diff --git a/package.json b/package.json
index afe80c0..b09adf5 100644
--- a/package.json
+++ b/package.json
@@ -7,7 +7,7 @@
     "dijit": "1.14.0",
     "dojo": "1.14.0",
     "dojo-util": "1.14.0",
-    "dojo-webpack-plugin": "2.7.7",
+    "dojo-webpack-plugin": "^2.7.8-0",
     "dojox": "1.14.0",
     "less": "2.7.1",
     "less-loader": "4.1.0",
diff --git a/webpack.config.js b/webpack.config.js
index b969705..4f3c192 100644
--- a/webpack.config.js
+++ b/webpack.config.js
@@ -48,6 +48,7 @@ module.exports = {
                        environment: {dojoRoot: "release"},     // used at run time for non-packed resources (e.g. blank.gif)
                        buildEnvironment: {dojoRoot: "node_modules"}, // used at build time
                        locales: ["en"],
+                       async: true,
                        noConsole: true
                }),
```

**Comment by cmdcolin** [comment]
Thanks for fixing :) I am trying to follow the development but is it possible to load AMD modules that were not available at build time using this technique e.g. the app is built, then someone wants to add a plugin later? I'd be really interested if that was possible

**Comment by chuckdumont** [comment]
Hi @cmdcolin.  So you're talking about adding a Dojo loader plugin to a Dojo app that has been built with webpack, and you don't want to rebuild it?  I'm not sure how that would work, but since dojo-webpack-plugin is a webpack plugin that runs at build time, I don't see how it could be of any help.

**Comment by cmdcolin** [comment]
The idea is to include some code via require that was not built originally in the webpack build. Previous iterations of our code were able to use dojo.require "packages" to point to them https://dojotoolkit.org/reference-guide/1.7/loader/amd.html#id8

I understand that webpack generally wants all dependencies being available at build time so this is sort of a strange request

**Comment by chuckdumont** [comment]
The problem with requiring modules not included in the build is that those modules can have their own dependencies which would also need to be loaded, so you would need a full blown AMD loader in order to be able to resolve and load them.  Webpack is not an AMD loader.  Webpack converts AMD modules to a format similar to CommonJS.  Loading AMD modules that were not included in the build is not something webpack supports directly.

You might want to look into using [d3-require](https://www.npmjs.com/package/d3-require) or something similar.  See the [post](https://github.com/OpenNTF/dojo-webpack-plugin/issues/183#issuecomment-426625495) by @GordonSmith

**Comment by cmdcolin** [comment]
Got it thanks for clarifying!

**Comment by whatitis1** [comment]
@chuckdumont Sorry for the delay. Just got a chance to test it out and all looks good so far. Thanks!

**Comment by chuckdumont** [comment]
Support for the async option is now available in v2.8.0.
